### PR TITLE
Compose to ACI mapping documentation

### PR DESCRIPTION
**Pull request by gtardif**
Signed-off-by: Guillaume Tardif <guillaume.tardif@docker.com>

**What I did**
Document ACI support for compose fields

**Related issue**
https://github.com/docker/compose-cli/issues/253

**(not mandatory) A picture of a cute animal, if possible in relation with what you did**

### CODE REVIEW

**Comment by gtardif** [comment]
cc @karolz-ms. 
You can also take a look at https://github.com/docker/compose-cli/blob/main/docs/aci-interation.md for higher level feature docs on the ACI integration.

File: docs/aci-compose-features.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+volumes:
+  mydata:
+    driver: azure_file
+    driver_opts:
+      share_name: myfileshare
+      storage_account_name: mystorageaccount
+```
+
+Short volume syntax is not allowed for ACI volumes, as it wa designed for local path when running local containers.
```
Comment (chris-crone) [inline docs/aci-compose-features.md:115]:
```suggestion
The short volume syntax is not allowed for ACI volumes, as it was designed for local path bind mounting when running local containers.
```

File: docs/aci-compose-features.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+volumes:
+  mydata:
+    driver: azure_file
+    driver_opts:
+      share_name: myfileshare
+      storage_account_name: mystorageaccount
+```
+
+Short volume syntax is not allowed for ACI volumes, as it wa designed for local path when running local containers.  
+A Compose file can defined several volumes, with different Azure file shares or storage accounts.
```
Comment (chris-crone) [inline docs/aci-compose-features.md:116]:
```suggestion
A Compose file can define several volumes, with different Azure file shares or storage accounts.
```

File: docs/aci-compose-features.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+Credentials for storage account will be automatically fetched at deployment time, using the Azure login to retrive storage account key for each storage account used.
```
Comment (chris-crone) [inline docs/aci-compose-features.md:118]:
```suggestion
Credentials for storage accounts will be automatically fetched at deployment time using the Azure login to retrieve the storage account key for each storage account used. 
```

File: docs/aci-compose-features.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+- __✓:__ Converts
+- __n:__ Not yet implemented
+- __x:__ Not applicable / no 1-1 conversion
+
+| Keys                           | Map|  Notes                                                       |
+|--------------------------------|----|--------------------------------------------------------------|
+| __Service__                    | ✓  |                                                              |
+| service.service.build          | x  |  Ignored. no image build support on ACI.                              |
+| service.cap_add, cap_drop      | x  |                                                              |
+| service.command                | ✓  |  Override container Command. on ACI, specifying command will currently also override image entrypoint if the image has an Entrypoint defined |
```
[the remark this answers is not in the document]
Comment (karolz-ms) [reply docs/aci-compose-features.md:22]: So the semantics here is really equivalent to entrypoint override, even though the API naming is about "overriding container command"?

File: docs/aci-compose-features.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+| service.configs                | n  |                                                              |
```
Comment (karolz-ms) [inline docs/aci-compose-features.md:23]: Should this be 'x'? AFAIK ACI does not support the config concept?

File: docs/aci-compose-features.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+| service.deploy.resources       | ✓  |  Restriction: ACI Resource limits cannot be greater than the sum of resource requests for all containers in the container group. This will allow one containers in the same container group to compete with resources. 
+| service.deploy.restart_policy  | ✓  |  One of: any - none - on-failure. Restriction: All services must have the ame restart policy. The entire ACI container group will be restarted if needed. | 
+| service.deploy.labels          | x  |  ACI does not have container-level labels. In addition, Azure tag are too restrictive to support Docker container labels.| 
+| service.devices                | x  | 
+| service.depends_on             | x  | 
+| service.dns                    | x  | 
+| service.dns_search             | x  | 
+| service.domainname             | ✓  |  Mapped to ACI DNSLabelName. Restriction: all services must specify the same domainname, if specified.
+| service.tmpfs                  | x  |  
+| service.entrypoint             | x  |  ACI allows to override Command, but not Entrypoint.
```
[the remark this answers is not in the document]
Comment (karolz-ms) [reply docs/aci-compose-features.md:42]: See my comment about `service.command` above

File: docs/aci-compose-features.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+| service.tmpfs                  | x  |  
+| service.entrypoint             | x  |  ACI allows to override Command, but not Entrypoint. 
+| service.env_file               | ✓  |  
+| service.environment            | ✓  |  
+| service.expose                 | x  |   
+| service.extends                | x  |  
+| service.external_links         | x  |  
+| service.extra_hosts            | x  |  
+| service.group_add              | x  |  
+| service.healthcheck            | x  |
```
Comment (karolz-ms) [inline docs/aci-compose-features.md:50]: Should this be a `n`? ACI does support liveness and readiness probes

File: docs/aci-compose-features.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+| service.hostname               | x  |  
+| service.image                  | ✓  |  Private images will be accessible if the user is logged in against the corresponding registry at deploy time. Users will be automatically logged in to Azure Container Registry from Azure login if possible.
+| service.isolation              | x  |  
+| service.labels                 | x  |  ACI does not have container-level labels. In addition, Azure tag are too restrictive to support Docker container labels.  |
+| service.links                  | x  |  
+| service.logging                | x  |  
+| service.network_mode           | x  |  
+| service.networks               | x  |  Communication between services is supported by defining mapping for each service in `/etc/hosts` in the container group. Each service can resolve names for other services and the resulting network calls will be redirected to `localhost`.
+| service.pid                    | x  |  
+| service.ports                  | ✓  |
```
Comment (karolz-ms) [inline docs/aci-compose-features.md:60]: It is probably worth mentioning that only symmetric port mapping is supported

File: docs/aci-compose-features.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+| service.image                  | ✓  |  Private images will be accessible if the user is logged in against the corresponding registry at deploy time. Users will be automatically logged in to Azure Container Registry from Azure login if possible.
+| service.isolation              | x  |  
+| service.labels                 | x  |  ACI does not have container-level labels. In addition, Azure tag are too restrictive to support Docker container labels.  |
+| service.links                  | x  |  
+| service.logging                | x  |  
+| service.network_mode           | x  |  
+| service.networks               | x  |  Communication between services is supported by defining mapping for each service in `/etc/hosts` in the container group. Each service can resolve names for other services and the resulting network calls will be redirected to `localhost`.
+| service.pid                    | x  |  
+| service.ports                  | ✓  |  
+| service.secrets                | ✓  |
```
Comment (karolz-ms) [inline docs/aci-compose-features.md:61]: Didn't know that secrets are supported today? @gtardif  can you point me to an example?

Comment (karolz-ms) [reply docs/aci-compose-features.md:61]: Probably just a typo? 😞

File: docs/aci-compose-features.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+| service.stop_grace_period      | x  |  
+| service.stop_signal            | x  |  
+| service.sysctls                | x  |  
+| service.ulimits                | x  |  
+| service.userns_mode            | x  |  
+| service.volumes                | ✓  |  Mapped to AZure File Shares. See #persistent_volumes
+| service.restart                | x  |  Replaced by service.deployment.restrt_policy                |
+|                                |    |                                                              |
+| __Volume__                     | x  |                                                              |
+| driver                         | ✓  |                                                              |
```
Comment (karolz-ms) [inline docs/aci-compose-features.md:72]: Suggestion: specify what the Azure File Share valid driver name and options are

File: docs/aci-compose-features.md (hunk 8 above)
Comment (karolz-ms) [reply docs/aci-compose-features.md:60]: I see that you talk about it below, but such important limitation deserves a comment here as well IMO

File: docs/aci-compose-features.md (hunk 10 above)
Comment (karolz-ms) [reply docs/aci-compose-features.md:72]: Oh, I see, there is more info below. How to add a comment with a link to the relevant paragraph?

File: docs/aci-compose-features.md [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+| service.cgroup_parent          | x  |  
+| service.container_name         | x  |  Service name is used as container name in ACI               |
+| service.credential_spec        | x  |                                                              |
+| service.deploy                 | ✓  |                                                              |
+| service.deploy.endpoint_mode   | x  |                                                              |
+| service.deploy.mode            | x  |                                                              |
+| service.deploy.replicas        | x  |  Only one replica is started for each service with the docker-ACI integration. | 
+| service.deploy.placement       | x  |  
+| service.deploy.update_config   | x  |  
+| service.deploy.resources       | ✓  |  Restriction: ACI Resource limits cannot be greater than the sum of resource requests for all containers in the container group. This will allow one containers in the same container group to compete with resources.
```
[the remark this answers is not in the document]
Comment (karolz-ms) [reply docs/aci-compose-features.md:33]: Is this where CPU and memory limits are specified? Perhaps an example would be helpful

File: docs/aci-compose-features.md (hunk 5 above)
Comment (gtardif) [reply docs/aci-compose-features.md]: updated

File: docs/aci-compose-features.md [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+Table bellow list supported Compose file fields with equivalent ACI when supported.
+
+__Glossary:__
+
+- __✓:__ Converts
+- __n:__ Not yet implemented
+- __x:__ Not applicable / no 1-1 conversion
+
+| Keys                           | Map|  Notes                                                       |
```
[the remark this answers is not in the document]
Comment (gtardif) [reply docs/aci-compose-features.md]: Actually, removing the last | at the end of lines makes this a lot easier to maintain, and displays nicely (https://github.com/gtardif/compose-cli/blob/doc_aci_compose/docs/aci-compose-features.md)

File: docs/aci-compose-features.md (hunk 9 above)
Comment (gtardif) [reply docs/aci-compose-features.md]: actually no, we do support them, we have not advertise them so far, the mounted folders are slightly different from a local compose app, so it's not quite portable. I've added that in the docs with a specific paragraph. Will look again to see if we can have a better folder mapping

Comment (karolz-ms) [reply docs/aci-compose-features.md]: That's awesome, thanks!

File: docs/aci-compose-features.md [hunk 13]
```diff
@@ -0,0 +1,175 @@
+# Compose - Azure Container Instances mapping
+
+This document outlines the conversion of an application defined in a Compose file to ACI objects.
+At a high-level, each Compose deployment is mapped to a single ACI container group. 
+Each service is mapped to a container in the container group. The Docker ACI integration provides does not allow scaling of services.
```
Comment (chris-crone) [inline docs/aci-compose-features.md:5]:
```suggestion
Each service is mapped to a container in the container group. The Docker ACI integration does not allow scaling of services.
```

**Review by chris-crone — APPROVED** [review]
LGTM

We should put this up on docs.docker.com too

**Pull request merged**